### PR TITLE
translations(rustc_session): migrates rustc_session to use SessionDiagnostic - Final

### PR DESCRIPTION
**Pull request by beowolx**
# Description
This is the final part of the rustc_session https://github.com/rust-lang/rust/issues/100717#issuecomment-1220279883.

Please only review this [commit](https://github.com/rust-lang/rust/pull/101266/commits/a54534703774bfb9fc344f61d511760a7c43fe94). The other ones are from the PR https://github.com/rust-lang/rust/pull/101041# that is not yet merged.

In this PR, we migrate the file `output.rs`

### CODE REVIEW

**Comment by LuisCardosoOliveira** [comment]
> 
> The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/8130556981?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/06cb86edbdfc6e8f894e1619b908117e18ccb29b/checks/8130556981/logs)
> 
> <details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>
> 
> ```plain
> 
> running 229 tests
> ..........i.....ii...................................................................... 88/229
> ..i.........................i...................iiiiiii......i...................iii.... 176/229
> Some tests failed in compiletest suite=run-make-fulldeps mode=run-make host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
> ...............................i...ii..F.............
> 
> ---- [run-make] src/test/run-make-fulldeps/weird-output-filenames stdout ----
> 
> error: make failed
> error: make failed
> status: exit status: 2
> command: "make"
> --- stdout -------------------------------
> cp foo.rs /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/weird-output-filenames/weird-output-filenames/.foo.rs
> LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/weird-output-filenames/weird-output-filenames:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/weird-output-filenames/weird-output-filenames -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/weird-output-filenames/weird-output-filenames  /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/weird-output-filenames/weird-output-filenames/.foo.rs 2>&1 \
>  | "/checkout/src/etc/cat-and-grep.sh" -e "invalid character.*in crate name:"
> [[[ begin stdout ]]]
> thread 'rustc' panicked at 'identifier: "session_invalid_character_in_create_name", attr: None, args: FluentArgs([]), errors: [ResolverError(Reference(Variable { id: "character" })), ResolverError(Reference(Variable { id: "crate_name" }))]', compiler/rustc_errors/src/translation.rs:91:17
> stack backtrace:
>    0:     0x7f6b41b5012c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hfc487dcf423f7dc5
>    1:     0x7f6b41bb8ca8 - core::fmt::write::h8b3d9a117ecda149
>    2:     0x7f6b41b409a1 - std::io::Write::write_fmt::hc1f8e5607fcae684
>    3:     0x7f6b41b5311e - std::panicking::default_hook::{{closure}}::h10b8ec90f5d1420b
>    4:     0x7f6b41b52de7 - std::panicking::default_hook::h077c4fc56990cab2
>    5:     0x7f6b424e2054 - rustc_driver[939608cf3f9c8401]::DEFAULT_HOOK::{closure#0}::{closure#0}
>    6:     0x7f6b41b538d1 - std::panicking::rust_panic_with_hook::h255d1fc3fdfe100d
>    7:     0x7f6b41b536f7 - std::panicking::begin_panic_handler::{{closure}}::hbe73b989bc6b0e39
>    8:     0x7f6b41b506a4 - std::sys_common::backtrace::__rust_end_short_backtrace::hb90e20907a7a19fd
>    9:     0x7f6b41b533c2 - rust_begin_unwind
>   10:     0x7f6b41b03e43 - core::panicking::panic_fmt::h206079c3bf5fa5e3
>   11:     0x7f6b451d8a6a - <rustc_errors[80a7ddfbd784b988]::emitter::EmitterWriter as rustc_errors[80a7ddfbd784b988]::translation::Translate>::translate_message
>   12:     0x7f6b451ce052 - <rustc_errors[80a7ddfbd784b988]::emitter::EmitterWriter>::emit_message_default
>   13:     0x7f6b451cbb72 - <rustc_errors[80a7ddfbd784b988]::emitter::EmitterWriter as rustc_errors[80a7ddfbd784b988]::emitter::Emitter>::emit_diagnostic
>   14:     0x7f6b452258f8 - <rustc_errors[80a7ddfbd784b988]::HandlerInner>::emit_diagnostic
>   15:     0x7f6b452215a6 - <rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed as rustc_errors[80a7ddfbd784b988]::diagnostic_builder::EmissionGuarantee>::diagnostic_builder_emit_producing_guarantee
>   16:     0x7f6b4514a21f - <rustc_session[acee3a4a0a70eadd]::parse::ParseSess>::emit_err::<rustc_session[acee3a4a0a70eadd]::errors::InvalidCharacterInCrateName>
>   17:     0x7f6b451a8a34 - rustc_session[acee3a4a0a70eadd]::output::validate_crate_name
>   18:     0x7f6b451a8866 - rustc_session[acee3a4a0a70eadd]::output::find_crate_name
>   19:     0x7f6b4261b544 - <rustc_interface[be3e3b91a6472b5b]::queries::Queries>::crate_name
>   20:     0x7f6b4261adca - <rustc_interface[be3e3b91a6472b5b]::queries::Queries>::register_plugins
>   21:     0x7f6b424e3792 - <rustc_interface[be3e3b91a6472b5b]::interface::Compiler>::enter::<rustc_driver[939608cf3f9c8401]::run_compiler::{closure#1}::{closure#2}, core[e1c75ed3987e2883]::result::Result<core[e1c75ed3987e2883]::option::Option<rustc_interface[be3e3b91a6472b5b]::queries::Linker>, rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>>
>   22:     0x7f6b424d0551 - rustc_span[6fde1357d18f6af6]::with_source_map::<core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>, rustc_interface[be3e3b91a6472b5b]::interface::create_compiler_and_run<core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>, rustc_driver[939608cf3f9c8401]::run_compiler::{closure#1}>::{closure#1}>
>   23:     0x7f6b424e6661 - rustc_interface[be3e3b91a6472b5b]::interface::create_compiler_and_run::<core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>, rustc_driver[939608cf3f9c8401]::run_compiler::{closure#1}>
>   24:     0x7f6b424cd3c2 - <scoped_tls[723eb8456d42f204]::ScopedKey<rustc_span[6fde1357d18f6af6]::SessionGlobals>>::set::<rustc_interface[be3e3b91a6472b5b]::interface::run_compiler<core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>, rustc_driver[939608cf3f9c8401]::run_compiler::{closure#1}>::{closure#0}, core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>>
>   25:     0x7f6b4254554f - std[3417fb0cb2b602d0]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[be3e3b91a6472b5b]::util::run_in_thread_pool_with_globals<rustc_interface[be3e3b91a6472b5b]::interface::run_compiler<core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>, rustc_driver[939608cf3f9c8401]::run_compiler::{closure#1}>::{closure#0}, core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>>::{closure#0}, core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>>
>   26:     0x7f6b424cd9ae - std[3417fb0cb2b602d0]::panicking::try::<core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>, core[e1c75ed3987e2883]::panic::unwind_safe::AssertUnwindSafe<<std[3417fb0cb2b602d0]::thread::Builder>::spawn_unchecked_<rustc_interface[be3e3b91a6472b5b]::util::run_in_thread_pool_with_globals<rustc_interface[be3e3b91a6472b5b]::interface::run_compiler<core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>, rustc_driver[939608cf3f9c8401]::run_compiler::{closure#1}>::{closure#0}, core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>>::{closure#0}, core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>>::{closure#1}::{closure#0}>>
>   27:     0x7f6b42548930 - <<std[3417fb0cb2b602d0]::thread::Builder>::spawn_unchecked_<rustc_interface[be3e3b91a6472b5b]::util::run_in_thread_pool_with_globals<rustc_interface[be3e3b91a6472b5b]::interface::run_compiler<core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>, rustc_driver[939608cf3f9c8401]::run_compiler::{closure#1}>::{closure#0}, core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>>::{closure#0}, core[e1c75ed3987e2883]::result::Result<(), rustc_errors[80a7ddfbd784b988]::ErrorGuaranteed>>::{closure#1} as core[e1c75ed3987e2883]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
>   28:     0x7f6b41b60645 - std::sys::unix::thread::Thread::new::thread_start::h1e1f1330b83c15f9
>   29:     0x7f6b418fcb43 - <unknown>
>   30:     0x7f6b4198ea00 - <unknown>
>   31:                0x0 - <unknown>
> error: internal compiler error: unexpected panic
> 
> note: the compiler unexpectedly panicked. this is a bug.
> 
> 
> note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md
> 
> note: rustc 1.65.0-nightly (e34cc674c 2022-09-01) running on x86_64-unknown-linux-gnu
> 
> query stack during panic:
> end of query stack
> 
> [[[ end stdout ]]]
> Error: cannot match: invalid character.*in crate name:
> --- stderr -------------------------------
> --- stderr -------------------------------
> make: *** [Makefile:5: all] Error 1
> 
> 
> 
> failures:
> ```
> 
> </details>
>         

I know why this test is failing. I forgot to add these params in the manually implementation of `SessionDiagnostic`. I'll correct it tonight.

**Comment by LuisCardosoOliveira** [comment]
I'm marking this PR as `draft` until I have the time to fix the test that is failling.

**Review by davidtwco — APPROVED** [review]
r=me after #101041 lands

**Comment by LuisCardosoOliveira** [comment]
Hey @davidtwco,

I have rebased this branch to update it with the last changes from https://github.com/rust-lang/rust/pull/101041#.

I still have the issue with `#[rustc_lint_diagnostics]`, so I had to annotate some of the functions with `allow`s as discussed before. You can check the diff [here](https://github.com/rust-lang/rust/compare/92dcfc914fc2e6837da597c8303261004a708e59..0c853b4ae484bc6b04d28db90cd1d8d7e1d16c78).

**Comment by LuisCardosoOliveira** [comment]
https://github.com/rust-lang/rust/pull/101041# landed! Go for review.

r? @davidtwco

File: compiler/rustc_session/src/errors.rs [hunk 1]
```diff
@@ -170,3 +170,62 @@ pub struct StackProtectorNotSupportedForTarget<'a> {
...
+}
+
+#[derive(SessionDiagnostic)]
+#[diag(session::crate_name_invalid)]
+pub struct CrateNameInvalid<'a> {
+    pub s: &'a str,
+}
+
+pub struct CrateNameEmpty {
+    pub span: Option<Span>,
```
Comment (davidtwco) [inline compiler/rustc_session/src/errors.rs:196]: Does `#[primary_span]` not work on an `Option<Span>` field?

Comment (LuisCardosoOliveira) [reply compiler/rustc_session/src/errors.rs:199]: It works and I though it didn't 😅 

Thanks David, I have just changed it. Here is the [diff](https://github.com/rust-lang/rust/compare/fbe3f99830cb4de9b814290ab1ce03d693c587c1..329d5014b6a773d819e6cd56d5930a204bed7983).

**Review by davidtwco — APPROVED** [review]

**Pull request merged**